### ISSUE TITLE
closing hotspot layer when clicking on image

### ISSUE THREAD

**Issue by kls0e** [naver/egjs-view360#310]
Hi,

thank you so much for egjs-view360, it is a remarkable piece of code and it is great fun using it. I have added some hot spots to a panoviewer panorama, and I would like to have the layer closed once the user clicks on the layered jpeg. I cannot seem to find the right way to modify the onclick="closeLayer(event);" or place it right. Currently there is 

`    <div class="layer" onclick="closeLayer(event);">
`
but this means that you have to click next to the layered jpeg to close the layer. And if I just append the onclick directive to the `<img>` it does not work. 

I am referring to this [implementation example](https://naver.github.io/egjs-view360/examples/panoviewer/etc/hotspot.html), any hints are appreciated. Thanks

**Comment by WoodNeck** [comment]
Hello, @kls0e!
I think the easiest solution for this issue is adding `pointer-events: none` CSS to an img element.

![1](https://user-images.githubusercontent.com/26213435/77600205-82be6400-6f4a-11ea-9754-1d1ab89c0ffa.gif)

But be careful that now the image element can't receive any click related events by setting that.

And the reason why the image element doesn't react even if it calls `closeLayer` is because of how that function is written.

![image](https://user-images.githubusercontent.com/26213435/77600517-8999a680-6f4b-11ea-83f7-79cb4ccb5e59.png)
So basically, you can remove the function's target checking to make an image onclick to work.

```diff
function closeLayer(e) {
- if (e.target === layer) {
    layer.style.display = "none";
- }
}
```

**Comment by kls0e** [comment]
Dear WoodNeck, 

many thanks for your elaborate help. I appreciate your support and both ways your provided are a perfect solution to my request. As I am making progress with my project, I have another three questions, perhaps I can ask them here, or should I create separate github issues? 

1. I'd like to make the hotspot glasses less bouncy when moving around, can I somehow reduce their vertical movement? (Im using panorama view)

2. How can I place a div with (different) text under each image? I'm building a virtual gallery and would like to put the artists' name under the corresponding image. Can the onclick-event of the hotspot symbols maybe transport two variables, one for the img url, one for the div content? Or how do I approach this?

3. is there an option to show the current yaw and pitch of the cursor? currently I place them using trial and error while changing the coordinates in the code. 

Best wishes, stay safe

Matthias

**Comment by WoodNeck** [comment]
I think it's just okay to continue on here as it's all hotspot-related questions :)

1. Well, I know what you mean with those buoyancy movements of hotspots.
I didn't actually made this demo but here's why:
The demo's background image is tilted about 10 degrees, so there's a compensation script for it to work.
Remove the following lines to remove that compensation.
```diff
function setHotspotOffset(hotspot, viewer) {
        var oyaw = viewer.getYaw();
        var opitch = viewer.getPitch();
        var yaw = parseFloat(hotspot.getAttribute("data-yaw"));
        var pitch = parseFloat(hotspot.getAttribute("data-pitch"));
        var deltaYaw = yaw - oyaw;
        var deltaPitch = pitch - opitch;

        if (deltaYaw < -180) {
            deltaYaw += 360;
        } else if (deltaYaw > 180) {
            deltaYaw -= 360;
        }
        if (Math.abs(deltaYaw) > 90) {
            hotspot.style.transform = "translate(-200px, 0px)";
            return;
        }
        var radYaw = toRadian(deltaYaw);
        var radPitch = toRadian(deltaPitch);

        var fov = viewer.getFov();
        var hfov = getHFov(fov);

        var rx = Math.tan(toRadian(hfov) / 2);
        var ry = Math.tan(toRadian(fov) / 2);


        var point = [
            Math.tan(-radYaw) / rx,
            Math.tan(-radPitch) / ry,
        ];

        // Image rotation compensation
        // The original image is about 10 degrees tilted.
-       point = point.map(function (p) {
-           return p * Math.cos(15 / 180 * Math.PI);
-       });
-       point[1] = rotate(point, deltaYaw > 0 ? -10 : 10)[1];

        // point[0] /= 1.05;
        var left = viewer._width / 2 + point[0] * viewer._width / 2;
        var top = viewer._height / 2 + point[1] * viewer._height / 2;

        hotspot.style.transform = "translate(" + left + "px, " + top + "px) translate(-50%, -50%)";
}
```

2. That's simple, add the div element inside the element with class "layer" and modify it just like the image element.

First change the page content like this
```diff
<div class="layer" onclick="closeLayer(event);">
  <img src="./img/book1.jpg">
+ <div class="layer-content"></div>
</div>
```

Grab that element, and modify it whatever you like on `openLayer()`.
```diff
var layer = document.querySelector(".layer");
+ var layerContent = document.querySelector(".layer-content");

-function openLayer(img) {
+function openLayer(img, content) {
  layer.querySelector("img").src = "./img/" + img;
  layer.style.display = "block";
+ layerContent.innerHTML = content;
}
```

And of course, you can transfer 2 variables at once.
```html
<div class="hotspot search" data-page="1" data-anchor-index="0" data-yaw="52" data-pitch="-14" onclick="openLayer('book1.jpg', 'This is second parameter')"></div>
<div class="hotspot search" data-page="1" data-anchor-index="1" data-yaw="-45" data-pitch="-18" onclick="openLayer('book2.jpg', 'You can put anything on here')"></div>
<div class="hotspot search" data-page="1" data-anchor-index="2" data-yaw="6" data-pitch="-17" onclick="openLayer('book3.jpg', 'Good Luck')"></div>
```

3. Actually there's one that we use for the local development, which looks like this.
![1](https://user-images.githubusercontent.com/26213435/77641807-ff2f6200-6f9f-11ea-8d0d-e6a22f1aabaf.gif)

There's source code on here: https://github.com/naver/egjs-view360/blob/master/test/manual/PanoViewer/PanoViewer.html#L459

The point is on the `render` function so check that out :)
You can use either gridView(that grid with a red circle on it), the file's on here: https://github.com/naver/egjs-view360/blob/master/test/manual/js/GridView.js
or just render the value direct as innerHTML like at the render function's Line 471~473

And also, stay safe you too :3

**Comment by kls0e** [comment]
this all is so beautiful and it works. I am grateful and your help means a lot to me. Will keep you posted when my project is done.

**Comment by WoodNeck** [comment]
Wow, that's really cool. I love it :)
Thanks for making a great product with our component.

I'll close this issue for now.
Feel free to reopen it whenever you like.
